Core/Core/Files/FileUploader.swift

```diff
@@ -0,0 +1,325 @@
+                let data = try encoder.encode(self)
+                return data.base64EncodedString()
+            } catch {
+                assertionFailure("failed to encode background identifier")
+                return ""
+            }
+        }
+    }
+
+    public let api: API
+    public let database: NSPersistentContainer
+    public var notificationManager: NotificationManager = .shared
+    var backgroundSession: URLSession!
+    lazy var backgroundAPI: API = URLSessionAPI(urlSession: backgroundSession)
+
+    /// Completion handler that should be called once all background tasks have finished.
+    ///
+    /// See UIApplicationDelegate.application(application:handleEventsForBackgroundURLSession:completionHandler:)
+    public var completionHandler: (() -> Void)?
+
+    private var listeners: [FileUploaderDelegate] = []
+
+    /// Initialize from background.
+    public convenience init?(backgroundSessionIdentifier id: String) {
+        guard let session = Session(string: id), let user = Keychain.entries.first(where: { $0 == session }) else {
+            return nil
+        }
+        let api = URLSessionAPI(accessToken: user.accessToken, actAsUserID: user.actAsUserID, baseURL: user.baseURL)
+        let database = NSPersistentContainer.create(appGroup: session.appGroup?.rawValue, session: user)
+        self.init(identifier: session.identifier, api: api, database: database, appGroup: session.appGroup)
+    }
+
+    /// Initialize from foreground.
+    public convenience init(bundleID: String, appGroup: AppGroup?, environment: AppEnvironment = .shared) {
+        guard let user = environment.currentSession else {
+            self.init(identifier: "\(bundleID).fileuploads", api: URLSessionAPI(), database: .create(), appGroup: appGroup)
+            return
+        }
+        let api = environment.api
+        let database = environment.database
+        let session = Session(bundleID: bundleID, appGroup: appGroup, userID: user.userID, baseURL: user.baseURL, actAsUserID: user.actAsUserID)
+        let identifier = session.identifier
+        self.init(identifier: identifier, api: api, database: database, appGroup: appGroup)
+    }
+
+    private init(identifier: String, api: API, database: NSPersistentContainer, appGroup: AppGroup?) {
+        Logger.shared.log("Creating file uploader with identifier \(identifier)")
+        self.api = api
+        self.database = database
+        let configuration = URLSessionConfiguration.background(withIdentifier: identifier)
+        configuration.sharedContainerIdentifier = appGroup?.rawValue
+        super.init()
+        backgroundSession = URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
+    }
+
+    public func upload(_ file: File, context: FileUploadContext, callback: @escaping (Error?) -> Void) {
+        Logger.shared.log()
+        let defaultError = NSError.instructureError(NSLocalizedString("Could not upload file", comment: ""))
```

Comment:
I've seen some `NSError`s and `enum: Error`s. Do we want to try to make that consistent, or do they have different uses?